modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java

```diff
@@ -710,27 +808,93 @@ public void visitBreak(SBreak userBreakNode, ScriptScope scriptScope) {
         Class<?> compoundType = scriptScope.hasDecoration(userAssignmentNode, CompoundType.class) ?
                 scriptScope.getDecoration(userAssignmentNode, CompoundType.class).getCompoundType() : null;
-        PainlessCast upcast = scriptScope.hasDecoration(userAssignmentNode, UpcastPainlessCast.class) ?
-                scriptScope.getDecoration(userAssignmentNode, UpcastPainlessCast.class).getUpcastPainlessCast() : null;
-        PainlessCast downcast = scriptScope.hasDecoration(userAssignmentNode, DowncastPainlessCast.class) ?
-                scriptScope.getDecoration(userAssignmentNode, DowncastPainlessCast.class).getDowncastPainlessCast() : null;
-
-        AssignmentNode irAssignmentNode = new AssignmentNode();
-        irAssignmentNode.setLeftNode((ExpressionNode)visit(userAssignmentNode.getLeftNode(), scriptScope));
-        irAssignmentNode.setRightNode(injectCast(userAssignmentNode.getRightNode(), scriptScope));
-        irAssignmentNode.setLocation(userAssignmentNode.getLocation());
-        irAssignmentNode.setExpressionType(scriptScope.getDecoration(userAssignmentNode, ValueType.class).getValueType());
-        irAssignmentNode.setCompoundType(compoundType);
-        irAssignmentNode.setPost(userAssignmentNode.postIfRead());
-        irAssignmentNode.setOperation(userAssignmentNode.getOperation());
-        irAssignmentNode.setRead(scriptScope.getCondition(userAssignmentNode, Read.class));
-        irAssignmentNode.setCat(scriptScope.getCondition(userAssignmentNode, Concatenate.class));
-        irAssignmentNode.setThere(upcast);
-        irAssignmentNode.setBack(downcast);
-
-        scriptScope.putDecoration(userAssignmentNode, new IRNodeDecoration(irAssignmentNode));
+
+        StoreNode irStoreNode;
+        ExpressionNode irValueNode = injectCast(userAssignmentNode.getRightNode(), scriptScope);
+
+        if (compoundType != null) {
+            scriptScope.setCondition(userAssignmentNode.getLeftNode(), Compound.class);
+            irStoreNode = (StoreNode)visit(userAssignmentNode.getLeftNode(), scriptScope);
+            BinaryMathNode irBinaryMathNode = (BinaryMathNode)irStoreNode.getChildNode();
+
+            PainlessCast downcast = scriptScope.hasDecoration(userAssignmentNode, DowncastPainlessCast.class) ?
+                    scriptScope.getDecoration(userAssignmentNode, DowncastPainlessCast.class).getDowncastPainlessCast() : null;
+
+            if (downcast == null) {
+                irBinaryMathNode.setExpressionType(irStoreNode.getStoreType());
+            } else {
+                CastNode irCastNode = new CastNode();
+                irCastNode.setLocation(irBinaryMathNode.getLocation());
+                irCastNode.setExpressionType(downcast.targetType);
+                irCastNode.setCast(downcast);
+                irCastNode.setChildNode(irBinaryMathNode);
+                irStoreNode.setChildNode(irCastNode);
+            }
+
+            if (read) {
```

Comment:
Bigger comment that says what we have to do to support it.

Reply:
Done.